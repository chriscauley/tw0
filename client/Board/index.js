import { range, pick, defaults } from 'lodash'

import Geo from '../Geo'
import vector from '../Geo/vector'
import { assert, floodFillPath, floodFillTeam } from '../utils'
import { newPiece } from '../piece/entity'

export default class Board {
  constructor({W, H, turn=0, ...options}) {
    this.entities = {
      piece: {},
      floor: {},
      team: {},
      path: {},
      wall: {},
      square: {}
    }
    this.options = defaults(options, {
      path: getDefaultPath,
      makeWalls: makeDefaultWalls
    })
    this.teams = [1, 2]
    this.cache = {}
    this.dirty = {
      team: true,
      path: true,
    }
    this._piece_id = 0
    Object.assign(this, { W, H })
    this.turn = turn
    this.geo = Geo(W, H)
    this.connectPath()
    this.recache()
  }

  setOne(type, index, value) {
    assert(type !== piece, 'cannot set piece with setOne')
    this.entities[type] = value
  }

  getOne = (type, index) => this.entities[type][index]
  getMany = (type, indexes) => indexes.map(i => this.getOne(type, i)).filter(i => i !== undefined)

  setPiece(index, piece) {
    this.dirty.team = true
    const old = this.entities.piece[index]
    assert(!old || old === piece, 'Pauli Exclusion error')

    if (piece.index !== undefined && piece.board) {
      // remove from old position
      piece.board.removePiece(piece)
    }

    piece.board = this
    piece.index = index
    this.entities.piece[index] = piece
  }

  removePiece(piece) {
    this.dirty.team = true
    delete this.entities.piece[piece.index]
    delete piece.board
  }

  getPieces() {
    return Object.values(this.entities.piece)
  }

  connectPath() {
    let { path, makeWalls } = this.options
    makeWalls(this)
    if (typeof path === 'function') {
      path = path(this)
    }

    this.start1 = path[0]
    this.start2 = path[path.length -1]
    const b = {}
    b[this.start1] = 1
    b[this.start2] = 2
    let last_index = path[0]
    path.slice(1).forEach(next_index => {
      const next_xy = this.geo.index2xy(next_index)
      const last_xy = this.geo.index2xy(last_index)
      const diff_xy = vector.subtract(next_xy, last_xy)
      assert(!(diff_xy[0] && diff_xy[1]), 'Path nodes should be orthogonal')
      const dindex = this.geo.xy2index(vector.sign(diff_xy))
      let i = 0
      while (last_index !== next_index) {
        this.entities.path[last_index] = dindex
        last_index += dindex
        i++
        if (i>1000) {
          throw "Path overflowing"
        }
      }
      this.entities.path[last_index] = dindex
    })

    this.geo.indexes.forEach(i => this.entities.square[i] = true)
  }

  newPiece(opts) {
    this.dirty.team = true
    const piece = newPiece(opts)
    piece.id = ++this._piece_id
    piece._turn = this.game ? this.game.turn : 0
    this.setPiece(piece.index, piece)
    return piece
  }

  recache() {
    if (this.dirty.path) {
      this.cache.path = floodFillPath(this)
    }
    if (this.dirty.team) {
      this.cache.team = {}
      const pieces = this.getPieces()
      this.teams.forEach((team) => {
        this.cache.team[team] = floodFillTeam(this, team)
      })
    }
    this.dirty = {}
  }
}

const getCorners = (board, d) => {
  const geo = board.geo
  return [geo.xy2index([d, d]), geo.xy2index([-d-1, -d])]
}

const getDefaultPath = (board) => {
  const { geo } = board
  const [start, end] = getCorners(board, 2)
  const b = {}
  const start_xy = geo.index2xy(start)
  const end_xy = geo.index2xy(end)
  const out = [start]
  if (start_xy[1] !== end_xy[1]) {
    const mid_x = start_xy[0] + Math.floor((end_xy[0] - start_xy[0])/2)
    out.push(geo.xy2index([mid_x, start_xy[1]]))
    out.push(geo.xy2index([mid_x, end_xy[1]]))
  }
  out.push(end)
  return out
}

const makeDefaultWalls = (board) => {
  const {geo, entities} = board
  const wall_indexes = geo.indexes.filter(i => {
    return (
      i < geo.W ||
        i > geo.AREA - geo.W ||
        i % geo.W === 0 ||
        i % geo.W === geo.W -1
    )
  })
  wall_indexes.forEach(i => entities.wall[i] = 'X')
}