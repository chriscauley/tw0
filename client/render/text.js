const LEGEND = {
  skeleton: 's',
  skull: 'k',
}

const _join = (strings) => {
  // combine several multi-line strings into one multiline string (horizontal concatenation)
  const result = []
  const splitted_strings = strings.map(s => s.split('\n'))
  splitted_strings[0].forEach((_, i) => {
    result.push(splitted_strings.map(lines => lines[i]).join('\t'))
  })
  return result.join('\n')
}

export default (board, layers, options={}) => {
  const { geo, entities } = board
  const {
    xy=[0,0],
    W=geo.W,
    H=geo.H,
  } = options
  const indexes = geo.slice(xy, W, H)
  return _join(layers.map(layer => renderLayer(board, layer, indexes)))
}

const getPieceId = (b, i) => b.entities.piece[i] && b.entities.piece[i].id
const getPieceIndex = (b, i) => b.entities.piece[i] && b.entities.piece[i].index
const getPieceDindex = (board, index) => {
  const piece = board.entities.piece[index]
  if (piece) {
    return board.geo._dindex2char[piece.dindex] || 'o'
  }
}

const getPieceType = (board, index) => {
  const piece = board.entities.piece[index]
  if (piece) {
    const { type, team } = piece
    const s = LEGEND[type] || '?'
    return team === 2 ? s.toUpperCase() : s
  }
}

const renderLayer = (board, layer, indexes) => {
  const getter = LAYERS[layer]
  const out = {}
  indexes.forEach((index) => {
    const wall = board.entities.wall[index]
    const path = board.entities.path[index]
    const value = getter(board, index)
    if (value) {
      out[index] = value
    } else if (wall) {
      out[index] = 'X'
    } else if (path) {
      out[index] = Math.abs(path) === 1 ? '-' : '|'
    } else {
      out[index] = '.'
    }
  })
  const title = layer.replace(/^piece_/,'').slice(0, board.geo.W).padEnd(board.geo.W)
  return board.geo.print(out, {title})
}

const LAYERS = {
  piece_index: getPieceIndex,
  piece_id: getPieceId,
  piece_dindex: getPieceDindex,
  piece_type: getPieceType,
}