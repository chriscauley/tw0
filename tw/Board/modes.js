import { sortBy } from 'lodash'
import vector from '../Geo/vector'

const { subtract, magnitude } = vector

// this is based off the same algorithm as @unrest/random.
// I was getting repeats so I squared seed and they went away
const M = 0x80000000 - 1 // 2**31 - 1;
const A = 1103515245
const rand = (seed) => (seed * seed * A) % M
rand.choice = (s, array) => array[rand(s) % array.length]

class Disco {
  constructor(board) {
    this.board = board
    board.entities.node[board.options.nodes[0]] = 1
    board.options.nodes.slice(1).forEach((i) => (board.entities.node[i] = 2))
  }
  tick() {
    const { board } = this
    if (!Object.values(board.entities.node).find((team) => team === 2)) {
      board.level++
      const pieces = [...board.parsed_pieces[2]]
      let i = 0
      while (pieces.length < board.level) {
        pieces.push(pieces[i%pieces.length])
        i++
      }
      // blue controls all the nodes, switch all but closest node to red and respawn red pieces
      const player_xy = board.geo.index2xy(board.player.index)
      const nodes = sortBy(
        Object.keys(board.entities.node).map((n) => parseInt(n)),
        (i) => magnitude(subtract(player_xy, board.geo.index2xy(i))),
      ).slice(1)

      const dindexes = board.geo.look('box', 0, 3, 1)

      nodes.forEach((node_index, i) => {
        board.entities.node[node_index] = 2
        let tries = 0
        while (pieces.length) {
          const index = node_index + rand.choice(board.turn * node_index + tries + i, dindexes)
          if (board.canMoveOn(index)) {
            board.newPiece({ team: 2, type: pieces.pop(), index })
          }
          tries++
          if (tries > 100) {
            console.error('failed to place peace after 100 tries')
          }
        }
      })
    }
  }
}

class None {
  constructor(board) {
    this.board = board
  }
  tick() {}
}
export default { disco: Disco, none: None }
