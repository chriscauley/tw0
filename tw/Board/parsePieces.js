export default (pieces) => {
  if (typeof pieces === 'string') {
    if (!pieces.includes('+')) {
      pieces += '+' + pieces
    }
    const out = { 1: [], 2: [] }
    pieces.split('+').forEach((s, i_team) => {
      const team = i_team + 1
      s.split(',').forEach((s2) => {
        if (!s2) {
          return
        }
        if (!s2.match(/^\d+x/)) {
          s2 = '1x' + s2
        }
        const [_, count, type] = s2.match(/^(\d+)x(.+)/)
        let i = parseInt(count)
        while (i--) {
          out[team].push(type)
        }
      })
    })
    pieces = out
  }
  return pieces
}
