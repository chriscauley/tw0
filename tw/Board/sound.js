import { range } from 'lodash'

const MAX_SOUND = 8

const age = (board) => {
  // every turn the sound ages one layer value, and sounds "above" MAX_SOUND die
  const entries = Object.entries(board.entities.sound)
  entries.forEach((il) => il[1]++)
  board.entities.sound = Object.fromEntries(entries.filter((il) => il[1] < MAX_SOUND))
}

const cache = (board) => {
  const cache = (board.cache.sound = {})
  const sound_levels = range(MAX_SOUND).map(() => [])
  Object.entries(board.entities.sound).forEach((il) => sound_levels[il[1]].push(parseInt(il[0])))

  sound_levels.forEach((indexes, level) => {
    // this "junk" next level means no need to check level every time
    const next_level = sound_levels[level + 1] || { push: () => {} }

    indexes.forEach((index) => {
      if (cache[index] === undefined && !board.getOne('wall', index)) {
        cache[index] = level
        board.geo.look('circle', index, 1, 1).forEach((index2) => next_level.push(index2))
      }
    })
  })
}

export default { age, cache, MAX_SOUND }
