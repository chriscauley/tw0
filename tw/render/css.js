import { range } from 'lodash'
import types from 'tw/piece/types'

const piece_map = {}
types.slugs.forEach(s => piece_map[s] = types[s].sprite)

const addHealth = (piece, out) => {
  const blood = piece.blood_armor
  const HEART = `sprite sprite-heart ${blood ? 'micro' : 'mini'}-sprite`
  const BLOOD_ARMOR = 'sprite sprite-blood_armor mini-sprite'
  const INVINCIBLE = 'sprite sprite-invincible'
  if (piece.invincible) {
    out.children.push(INVINCIBLE)
  }
  if (piece.health > 100) {
    // show nothing
  } else if (piece.health > 4) {
    out.children.push(HEART)
    out.children.push('fa fa-question text-white mini-sprite')
  } else if (piece.max_health > 1) {
    range(piece.health).forEach(() => out.children.push(HEART))
  }
  if (blood) {
    range(blood).forEach(() => out.children.push(BLOOD_ARMOR))
  }
}

export const extra_getters = {
  off: () => ({}),
  sound: (b) => b.entities.sound,
  sound_cache: (b) => b.cache.sound,
  path: (b) => b.entities.path,
}

const team_extras = ['id', 'value', 'index', 'dindex']
;[1, 2].forEach((team) =>
  team_extras.forEach(
    (attr) =>
      (extra_getters[`team${team}_${attr}`] = (b) =>
        Object.fromEntries(
          Object.entries(b.cache.team[team]).map(([index, cache]) => [index, cache[attr]]),
        )),
  ),
)

export default (board, options = {}) => {
  const extra_layer = (extra_getters[options.extra] || extra_getters.off)(board)
  const { geo } = board
  const { xy0 = [0, 0], W = geo.W, H = geo.H, _empty, _highlight = [] } = options
  const items = []
  const indexes = geo.slice(xy0, W, H)
  const pieces = {}
  const index2xy = (i) => {
    const xy = geo.index2xy(i)
    return [xy[0] - xy0[0], xy[1] - xy0[1]]
  }

  const css = {
    index: (i) => {
      const xy = index2xy(i)
      return ` x-${xy[0]} y-${xy[1]}`
    },
    wall: (i, v) => ({
      className: `sprite ${css.index(i)} wall-${v} sprite-wall${((i + 151) % 127) % 5}`,
      id: `wall-${i}`,
      index: i,
    }),
    sound: (i, v) => ({
      id: `sound-${i}`,
      className: `sound sprite ${css.index(i)}`,
      text: v,
      index: i,
    }),
    square: (i) => {
      const xy = geo.index2xy(i)
      const n = ((xy[0] % 2) + (xy[1] % 2)) % 2
      return {
        className: `sprite square sprite-floor${n} ${css.index(i)}`,
        id: `square-${i}`,
        index: i,
      }
    },
    piece: (p) => ({
      // no index because this will be set several times
      className: `sprite piece ${p.wait !== undefined ? 'wait-' + p.wait : ''} team-${p.team}`,
      id: `piece-${p.id}`,
      children: [`sprite sprite-${piece_map[p.type]} ${p.awake ? 'awake' : ''}`],
      steps: [],
      index: p.index,
    }),
    node: (i, v) => ({
      className: `sprite sprite-node-${v} node ${css.index(i)}`,
      id: `node-${i}`,
      index: i,
    }),
  }

  const spritePiece = (piece) => {
    const out = css.piece(piece)
    out._base = out.className
    out.className += css.index(piece._indexes[0])
    const di_name = geo._dindex2name[piece.dindex]
    piece._indexes.forEach((i, n) => {
      let cls = out._base + css.index(i)
      const attack = animations[i].filter((a) => a.type === 'attack')[0]
      if (attack) {
        cls += ' attack-' + di_name
      }
      if (attack || n > 0) {
        out.steps.push(cls)
      }
    })
    const last = out.steps[out.steps.length - 1]
    if (last && last.includes(' attack-')) {
      out.steps.push(last.replace(/ attack-./, ''))
    }
    if (out.steps.length) {
      out.steps[out.steps.length - 1] += ` dindex-${di_name}`
    } else {
      out.className += ` dindex-${di_name}`
    }
    out.piece = piece
    pieces[piece.id] = piece
    items.push(out)
    addHealth(piece, out)
    return out
  }

  const animations = {}
  indexes.forEach((index) => {
    animations[index] = board.animations[index] || []
  })

  board.getPieces().forEach(spritePiece)
  indexes.forEach((index) => {
    // dead pieces aren't in board.getPieces()
    animations[index].forEach((a) => {
      if (a.type === 'death') {
        const item = spritePiece(a.piece)
        const last = item.steps[item.steps.length - 1]
        item.steps.push(last + ' is_dead')
      }
    })
  })

  indexes.forEach((index) => {
    const wall = board.getOne('wall', index)
    if (wall) {
      items.push(css.wall(index, wall))
    } else if (board.getOne('square', index)) {
      items.push(css.square(index))
      const extra_value = extra_layer[index]
      if (extra_value !== undefined) {
        items.push(css.sound(index, extra_value))
      }
    }
  })

  board.options.nodes.forEach((index) => {
    items.push(css.node(index, board.entities.node[index]))
  })

  return {
    boardClass: `board W-${W} H-${H}`,
    items,
  }
}
