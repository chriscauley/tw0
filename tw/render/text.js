import { assert } from '../utils'

const LEGEND = {
  skeleton: 's',
  skull: 'k',
}

const _join = (strings) => {
  // combine several multi-line strings into one multiline string (horizontal concatenation)
  const result = []
  const splitted_strings = strings.map(s => s.split('\n'))
  splitted_strings[0].forEach((_, i) => {
    result.push(splitted_strings.map(lines => lines[i]).join('\t'))
  })
  return result.join('\n')
}

export default (board, options={}) => {
  const { geo, entities } = board
  const {
    xy=[0,0],
    W=geo.W,
    H=geo.H,
    extra_layers=[],
    layer='piece_type'
  } = options
  const indexes = geo.slice(xy, W, H)
  const main = renderLayer(board, layer, indexes)
  const title = layer.replace(/^piece_/,'').slice(0, board.geo.W).padEnd(board.geo.W)
  const extras = []
  indexes.forEach((index) => {
    const [x,y] = geo.index2xy(index)
    const values = []
    extras[y] = extras[y] || ''
    extra_layers.forEach(layer => {
      const value = LAYERS[layer](board, index)
      if (value!==undefined) {
        values.push(value)
      }
    })
    if (values.length) {
      const joiner = extras[y] ? '\t' : ''
      extras[y] += `${index}:${values.join(',')}\t`
    }
  })
  return geo.print(main, { title, extras})
}

const getPieceId = (b, i) => b.entities.piece[i] && b.entities.piece[i].id
const getPieceIndex = (b, i) => b.entities.piece[i] && b.entities.piece[i].index
const getPieceDindex = (board, index) => {
  const piece = board.entities.piece[index]
  if (piece) {
    return board.geo._dindex2name[piece.dindex] || 'o'
  }
}

const getPieceType = (board, index) => {
  const piece = board.entities.piece[index]
  if (piece) {
    const { type, team } = piece
    const s = LEGEND[type] || '?'
    return team === 2 ? s.toUpperCase() : s
  }
}

const renderLayer = (board, layer, indexes) => {
  const getter = LAYERS[layer]
  assert(getter, 'You must specify a layer')
  const out = {}
  indexes.forEach((index) => {
    const wall = board.entities.wall[index]
    const path = board.entities.path[index]
    const value = getter(board, index)
    if (value) {
      out[index] = value
    } else if (wall) {
      out[index] = 'X'
    } else if (path) {
      out[index] = Math.abs(path) === 1 ? '-' : '|'
    } else {
      out[index] = '.'
    }
  })
  return out
}

const LAYERS = {
  piece_index: getPieceIndex,
  piece_id: getPieceId,
  piece_dindex: getPieceDindex,
  piece_type: getPieceType,
}