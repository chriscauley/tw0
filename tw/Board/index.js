import { defaults, range } from 'lodash'
import assert from 'assert'

import { canMoveOn } from '../piece/lib'
import Game from '../Game'
import Geo from '../Geo'
import vector from '../Geo/vector'
import { floodFillPath, floodFillTeam } from '../utils'
import { newPiece } from '../piece/entity'
import sound from './sound'
import modes from './modes'

export const parsePieces = (pieces) => {
  if (Array.isArray(pieces)) {
    throw 'Depracated: pieces should be object, not array'
  }
  if (typeof pieces === 'string') {
    if (!pieces.includes('+')) {
      pieces += '+' + pieces
    }
    const out = { 1: [], 2: [] }
    pieces.split('+').forEach((s, i_team) => {
      const team = i_team + 1
      s.split(',').forEach((s2) => {
        if (!s2) {
          return
        }
        if (!s2.match(/^\d+x/)) {
          s2 = '1x' + s2
        }
        const [_, count, type] = s2.match(/^(\d+)x(.+)/)
        let i = parseInt(count)
        while (i--) {
          out[team].push([type])
        }
      })
    })
    pieces = out
  }
  return pieces
}

export default class Board {
  constructor({ W, H, turn = 0, ...options }) {
    window.b = this
    this.entities = {
      piece: {},
      floor: {},
      team: {},
      path: {},
      wall: {},
      square: {},
      sound: {},
      node: {},
    }
    this.options = defaults(options, {
      path: getDefaultPath,
      makeWalls: makeDefaultWalls,
      nodes: [],
      mode: 'none',
      level: 1,
    })
    this.level = this.options.level
    this.teams = [1, 2]
    this.cache = {}
    this.dirty = {
      team: true,
      path: true,
    }
    this.pieces = []
    Object.assign(this, { W, H })
    this.turn = turn
    this.geo = Geo(W, H)
    this.dindex = this.geo.dindexes[0]
    this.connectPath()
    this.recache()
    sound.cache(this)
    this.game = options.game || new Game({ board: this, id: this.options.id })
    options.player && this.addPlayer(options.player)
    this.mode = new modes[this.options.mode](this)
    this.quickAddPieces(options.pieces)
    this.animations = {}
    this.recache()
  }

  getOne = (type, index) => this.entities[type][index]

  getMany = (type, indexes) =>
    indexes.map((i) => this.getOne(type, i)).filter((i) => i !== undefined)

  setOne(type, index, value) {
    assert(type !== 'piece', 'Use board.setPiece, not board.setOne("piece", ...)')
    this.entities[type] = value
  }

  getPiece = (index) => this.entities.piece[index]

  setPiece(index, piece) {
    this.dirty.team = true // TOOD I believe recaching is done every step in game so this is redundant
    const old = this.entities.piece[index]
    assert(!old || old === piece, 'Pauli Exclusion error')

    if (piece.index !== undefined && piece.board) {
      // remove from old position
      piece.board.removePiece(piece)
    }

    piece.board = this
    piece.index = index
    piece._indexes.push(index)

    const node_team = this.entities.node[index]
    if (node_team !== undefined && node_team !== piece.team) {
      this.entities.node[index] = piece.team
    }
    this.entities.piece[index] = piece
  }

  removePiece(piece) {
    this.dirty.team = true
    delete this.entities.piece[piece.index]
  }

  getPieces() {
    return Object.values(this.entities.piece)
  }

  // TODO this could be tw/Board/path.build and tw/Board/wall.build
  connectPath() {
    this.entities.node = {}
    this.entities.path = {}
    this.geo.indexes.forEach((i) => (this.entities.square[i] = true))
    const { makeWalls, walls } = this.options

    // walls generated by makeWalls are not stored in options.walls
    makeWalls(this)
    Object.assign(this.entities.wall, walls)

    if (this.options.nodes.length === 0) {
      this.options.nodes = getCorners(this, 3)
    }
    const nodes = this.options.nodes.slice() // clone because we might add one to it
    const first_xy = this.geo.index2xy(nodes[0])
    const last_xy = this.geo.index2xy(nodes[nodes.length - 1])
    this.start1 = nodes[0]
    this.start2 = nodes[nodes.length - 1]
    const half = Math.floor(nodes.length / 2)
    nodes.forEach((index, i) => {
      this.entities.node[index] = i >= half ? 2 : 1
    })
    if (nodes.length > 2 && vector.magnitude(vector.subtract(first_xy, last_xy)) < 8) {
      // nodes wrap around board
      nodes.push(nodes[0])
      this.start2 = nodes[half]
    }
    let last_node = nodes[0]
    // TODO this will break if not orthogonal
    // probably need to insert extra nodes into the temporary list of nodes
    nodes.slice(1).forEach((node) => {
      const xy1 = this.geo.index2xy(last_node)
      const xy2 = this.geo.index2xy(node)
      const dindex = this.geo.floorDindex(last_node - node)
      range(vector.magnitude(vector.subtract(xy1, xy2))).forEach((i_step) => {
        const index = node + i_step * dindex
        this.entities.path[index] = dindex
      })
      last_node = node
    })
  }

  newPiece(opts) {
    this.dirty.team = true
    const piece = newPiece(opts)
    piece.board = this
    piece.id = this.pieces.length + 1
    this.pieces.push(piece)
    piece._turn = this.game ? this.game.turn : 0
    piece._indexes = [] // used in renderer
    this.setPiece(piece.index, piece)
    return piece
  }

  canMoveOn(index) {
    // TODO move canMoveOn into board
    return canMoveOn(this, index)
  }

  getTeamSpawn(team) {
    const start = this['start' + team]
    const team_di = team === 1 ? 1 : -1
    const indexes = [start].concat(this.geo.look('circle', start, 2, team_di))
    const index = indexes.find((i) => canMoveOn(this, i))
    if (index === undefined) {
      throw `cannot place piece for team=${team}`
    }
    return index
  }

  quickAddPieces(pieces = {}) {
    pieces = parsePieces(pieces)
    Object.entries(pieces).forEach(([team, piece_set]) => {
      team = parseInt(team)
      piece_set.forEach(([type, index]) => {
        const options = { type, team, index }
        if (index === undefined) {
          options.index = this.getTeamSpawn(team)
        }
        this.newPiece(options)
      })
    })
  }

  getPathDindex(index, team) {
    return (team === 1 ? 1 : -1) * this.cache.path[index].dindex
  }

  recache() {
    if (this.dirty.path) {
      this.cache.path = floodFillPath(this)
    }
    if (this.dirty.team) {
      this.cache.team = {}
      this.teams.forEach((team) => {
        this.cache.team[team] = floodFillTeam(this, team)
      })
    }
    this.dirty = {}
  }

  animate(animation) {
    const { index, type } = animation
    this.animations[index] = this.animations[index] || []
    this.animations[index].push(animation)
    if (type === 'attack') {
      this.entities.sound[index] = 0
    }
  }

  // TODO this should be on game if we want to do multi board (not sure if I want to)
  addPlayer(team) {
    if (team) {
      const index = this.getTeamSpawn(team)
      this.player = this.newPiece({ team, type: 'warrior', player: true, index })
    }
  }

  startTurn() {
    sound.age(this)
    this.animations = {}
    this.getPieces().forEach((p) => (p._indexes = [p.index]))
  }
  cacheSound() {
    sound.cache(this)
  }
}

export const getCorners = (board, d) => {
  const geo = board.geo
  const x = geo.W < d * 2 ? Math.floor(geo.W / 2) : d
  const y = geo.H < d * 2 ? Math.floor(geo.H / 2) : d
  return [geo.xy2index([x, y]), geo.xy2index([-x - 1, -y])]
}

const getDefaultPath = (board) => {
  const { geo } = board
  const [start, end] = getCorners(board, 3)
  const start_xy = geo.index2xy(start)
  const end_xy = geo.index2xy(end)
  const out = [start]
  if (start_xy[0] !== end_xy[0] && start_xy[1] !== end_xy[1]) {
    const mid_x = start_xy[0] + Math.floor((end_xy[0] - start_xy[0]) / 2)
    out.push(geo.xy2index([mid_x, start_xy[1]]))
    out.push(geo.xy2index([mid_x, end_xy[1]]))
  }
  out.push(end)
  return out
}

const makeDefaultWalls = (board) => {
  const { geo, entities } = board
  const wall_indexes = geo.indexes.filter((i) => {
    return i < geo.W || i > geo.AREA - geo.W || i % geo.W === 0 || i % geo.W === geo.W - 1
  })
  wall_indexes.forEach((i) => (entities.wall[i] = 'X'))
}
