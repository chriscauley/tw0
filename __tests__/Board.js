import Board from '../tw/Board'
import render from '../tw/render/text'

const snap = (board) => {
  const result = render(board, 'piece_type', { extra_layers: ['piece_dindex'] })
  if (process.__UR && process.__UR.V) {
    console.log(result) // eslint-disable-line
  }
  expect(result).toMatchSnapshot()
}

test('2-skull', () => {
  const board = new Board({ W: 8, H: 3, pieces: 'skull' })
  snap(board)
  board.game.nextTurn()
  snap(board)
})

test('9-skull', () => {
  const pieces = [
    [1, 'skull', -6],
    [2, 'skull', -5],
    [1, 'skull', -4],
    [2, 'skull', -1],
    [1, 'skull', 0],
    [2, 'skull', 1],
    [1, 'skull', 4],
    [2, 'skull', 5],
    [1, 'skull', 6],
  ]
  const board = new Board({ W: 5, H: 5, path: [12], pieces })
  snap(board)
  board.game.nextTurn()
  snap(board)
})

test('getDefaultPath', () => {
  const tests = [
    { W: 5, H: 5 },
    { W: 5, H: 10 },
    { W: 10, H: 5 },
    { W: 10, H: 10 },
  ]
  tests.forEach((options) => snap(new Board(options)))
})

test('bonetar-conflicting-attack-move', () => {
  const pieces = [
    [1, 'skeleton', -5],
    [1, 'skeleton', -1],
    [2, 'bonetar', 1],
    [2, 'bonetar', 5],
  ]
  const board = new Board({ W: 5, H: 5, path: [12], pieces })
  snap(board)
  board.game.nextTurn()
  board.game.nextTurn()
  snap(board)
})

test('colliding skulls', () => {
  const board = new Board({ W: 11, H: 5, pieces: 'skull' })
  snap(board)
  board.game.nextTurn()
  board.game.nextTurn()
  snap(board)
})
