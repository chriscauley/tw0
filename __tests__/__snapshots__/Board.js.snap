// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`2-skull 1`] = `
"type    
XXXXXXXX	
X k--K X	
XXXXXXXX	"
`;

exports[`2-skull 2`] = `
"type    
XXXXXXXX	
X -kK- X	
XXXXXXXX	"
`;

exports[`2-skull 3`] = `
"type    
XXXXXXXX	
X ---- X	
XXXXXXXX	"
`;

exports[`2-skull 4`] = `
"type    
XXXXXXXX	
X ---- X	
XXXXXXXX	"
`;

exports[`9-skull 1`] = `
"type 
XXXXX	
XkKkX	
XKkKX	
XkKkX	
XXXXX	"
`;

exports[`9-skull 2`] = `
"type 
XXXXX	
X   X	
X   X	
Xk  X	
XXXXX	"
`;

exports[`getDefaultPath 1`] = `
"type 
XXXXX	
X   X	
X   X	
X   X	
XXXXX	"
`;

exports[`getDefaultPath 2`] = `
"type 
XXXXX	
X   X	
X | X	
X | X	
X | X	
X | X	
X | X	
X | X	
X   X	
XXXXX	"
`;

exports[`getDefaultPath 3`] = `
"type      
XXXXXXXXXX	
X        X	
X ------ X	
X        X	
XXXXXXXXXX	"
`;

exports[`getDefaultPath 4`] = `
"type      
XXXXXXXXXX	
X        X	
X --|    X	
X   |    X	
X   |    X	
X   |    X	
X   |    X	
X   ---- X	
X        X	
XXXXXXXXXX	"
`;
